src/app/txListener.js

```diff
@@ -0,0 +1,267 @@
+'use strict'
+var async = require('async')
+var ethJSABI = require('ethereumjs-abi')
+var ethJSUtil = require('ethereumjs-util')
+var EventManager = require('ethereum-remix').lib.EventManager
+var remix = require('ethereum-remix')
+var codeUtil = remix.util.code
+var Web3VMProvider = remix.web3.web3VMProvider
+
+/**
+  * poll web3 each 2s if web3
+  * listen on transaction executed event if VM
+  * attention: blocks returned by the event `newBlock` have slightly different json properties whether web3 or the VM is used
+  * trigger 'newBlock'
+  *
+  */
+class TxListener {
+  constructor (opt) {
+    this.event = new EventManager()
+    this._appAPI = opt.appAPI
+    this._web3VMProvider = new Web3VMProvider() // TODO this should maybe be put in app.js
+    this._web3VMProvider.setVM(opt.appAPI.vm())
+    this._resolvedTransactions = {}
+    this._resolvedContracts = {}
+    this.init()
+    opt.appEvent.executionContext.register('contextChanged', (context) => {
+      if (this.loopId) {
+        this.startListening(context)
+      }
+    })
+    opt.appEvent.udapp.register('transactionExecuted', (to, data, lookupOnly, txResult) => {
+      if (this.loopId && this._appAPI.isVM()) {
+        this._web3VMProvider.getTransaction(txResult.transactionHash, (error, tx) => {
+          if (error) return console.log(error)
+          this._newBlock({
+            type: 'VM',
+            number: -1,
+            transactions: [tx]
+          })
+        })
+      }
+    })
+  }
+
+  /**
+    * reset recorded transactions
+    */
+  init () {
```

Comment:
The `init` method looks like all of this could be already happen in the `constructor`.
So maybe it's not necessary to have the `init` method?